teleduck/src/teleduck/server.py

```
@@ -174,16 +173,26 @@ def run_server(
+        Checkpoints and closes DuckDB exactly once; safe to call repeatedly.
+        """
+        if shutdown_state["done"]:
+            return
+
+        shutdown_state["done"] = True
         try:
             duckdb_con.execute("CHECKPOINT")
             duckdb_con.close()
-            logging.info("atexit: database checkpointed and closed.")
+            logging.info("Database checkpointed and closed.")
         except Exception as exc:
```

Ruff (0.15.15)

[warning] Do not catch blind exception: `Exception`

(BLE001)